actionpack/lib/action_controller/metal/strong_parameters.rb

```diff
@@ -235,31 +248,86 @@ def ==(other)
     end
 
     # Returns a safe <tt>ActiveSupport::HashWithIndifferentAccess</tt>
-    # representation of this parameter with all unpermitted keys removed.
+    # representation of the parameters with all unpermitted keys removed.
     #
     #   params = ActionController::Parameters.new({
-    #     name: 'Senjougahara Hitagi',
-    #     oddity: 'Heavy stone crab'
+    #     name: "Senjougahara Hitagi",
+    #     oddity: "Heavy stone crab"
     #   })
-    #   params.to_h # => {}
+    #   params.to_h
+    #   # => ActionController::UnfilteredParameters: unable to convert unfiltered parameters to hash
     #
     #   safe_params = params.permit(:name)
     #   safe_params.to_h # => {"name"=>"Senjougahara Hitagi"}
     def to_h
       if permitted?
         convert_parameters_to_hashes(@parameters, :to_h)
+      elsif self.class.raise_on_unpermitted_parameters
+        raise UnfilteredParameters
       else
         slice(*self.class.always_permitted_parameters).permit!.to_h
       end
     end
 
+    # Returns a safe <tt>Hash</tt> representation of the parameters
+    # with all unpermitted keys removed.
+    #
+    #   params = ActionController::Parameters.new({
+    #     name: "Senjougahara Hitagi",
+    #     oddity: "Heavy stone crab"
+    #   })
+    #   params.to_hash
+    #   # => ActionController::UnfilteredParameters: unable to convert unfiltered parameters to hash
+    #
+    #   safe_params = params.permit(:name)
+    #   safe_params.to_hash # => {"name"=>"Senjougahara Hitagi"}
+    def to_hash
+      if self.class.raise_on_unpermitted_parameters
+        to_h.to_hash
+      else
+        @parameters.to_hash
```

Comment:
Yeah.. I actually meant "what to_h.to_hash would do if the config were enabled", but that [by design] isn't compatible either.

At minimum I think we should keep a deprecation here: we can't deprecate `to_h` with the config disabled, because we can't deprecate in a point release... but we can deprecate `to_hash` with the config disabled, because it's already showing a deprecation warning: we'd just be changing the message and thus the advice (to "[read about and then] turn on this config").

I'm still worried that even with that, as soon as `respond_to?(:to_hash)` becomes true, a lot of code will start treating this object differently (calling `to_hash` where it currently does not). Maybe we should special-case `:to_hash` in `respond_to?`, and lie about it when the config setting is off?